Прохожу спринт 17

EntityVadim opened the pull request:
Дорогой ревью, прошу тебя внести комментарии по коду, может, где-то, что-то стоит поменять, улучшить и т.д. или ссылки на полезные статьи, как я могу изменить свой код, может какие-то лайфхаки крутые подскажешь.
Буду тебе очень сильно в этом благодарен!

EntityVadim committed:
- Переместил файл с коре дата в отдельную папку, для лучшей читаемости в коде.

EntityVadim committed:
- Выполнил:
Самостоятельная задача
  - Добавте файл локализации Localizable.strings к трекеру привычек.
  - Перенесите все строковые литералы в Localizable.string.
  - Замените все упоминания строковых литералов на локализованный вариант с помощью NSLocalizedString.
  - Для плюрализации фразы «N дней» добавьте фразы в Localizable.stringsdict и настройте формы слова.

EntityVadim committed:
- Доделал задание:
  - Для плюрализации фразы «N дней» добавьте фразы в Localizable.stringsdict и настройте формы слова.

EntityVadim committed:
- Организовал код.

EntityVadim committed:
- Настроил приложение под темную тему;
- Мелкие исправления по коду;
- Настроил положение таблицы выбора дней недели;
- Убрал лишний разделитель в теблице выбора дней недели.

EntityVadim committed:
- Изменил цвета онбординга.

EntityVadim committed:
- Мелкие исправления по коду.

EntityVadim committed:
- Мелкие исправления по коду.

EntityVadim committed:
- Добавил поиск (фильтрация по названиям);
- Мелкие исправления по коду связанные с темной темой.

EntityVadim committed:
- Улучшил читаемость кода.

EntityVadim committed:
- Организовал код.

EntityVadim committed:
- Просил коммит.

EntityVadim committed:
- Реализовал "Контекстное меню".

EntityVadim committed:
- Исправил баг показа категорий без трекеров.

EntityVadim committed:
- Сделал отображение булавки при закрепление трекера и ее скрытия, после открепления трекера.

EntityVadim committed:
- Реализовал алерт удаления трекера.

EntityVadim committed:
- Переделал код для редактирования трекера.

EntityVadim committed:
- Реализовал подтягивания данных трекера (расписание и категории);
- Исправил баг с закрепление трекера.

EntityVadim committed:
- Исправил грубый баг показа кнопки расписания трекера у которого нет расписания.

EntityVadim committed:
- Организовал код.

EntityVadim committed:
- Настроил лейбл с днями.

EntityVadim committed:
- Создал экран "Фильтра"
- Сделал локализацию приложения.

EntityVadim committed:
- Мелкое исправление в локализации.

EntityVadim committed:
- datePicker установил текущую локаль устройства.

EntityVadim committed:
- Сверстал экран фильтры по макету.

EntityVadim committed:
- Мелкие исправления по локализации;
- Локализировал кнопку "Фильтры"

EntityVadim committed:
- Исправил баг с редактированием нерегулярного события.
- Мелкие изменения по коду.

EntityVadim committed:
- Исправил баг.

EntityVadim committed:
- Добавил счетчик дней на экране редактирования.

EntityVadim committed:
- Просил коммит.

EntityVadim committed:
- Контекстное меню теперь теперь взаимодействует с верхней частью ячейки, а не с целой ячейкой.

EntityVadim committed:
- Реализовал удаление категории "Закрепленные", если в нем нет трекеров.

EntityVadim committed:
- Просил коммит.

EntityVadim committed:
- Реализовал фильтрацию трекеров.

EntityVadim committed:
- Просил коммит.

EntityVadim committed:
- Реализовал логику по показу трекеров в выбраный день;
- Мелкие исправления.

EntityVadim committed:
- Сделал возможность сохранения фильтра при переходе на другую дату.

EntityVadim committed:
- Добавил возможность скрытия клавиатуры в поиске.

EntityVadim committed:
- Добавил заглушку;
- Скрыл кнопку "Фильтры" при отсутсвии трекеров.

EntityVadim committed:
- Применил заглушку при отрицательном поиске.

EntityVadim committed:
- Дополнил локализацию;
- Применил заглушку, если фильтр не дал результатов.

EntityVadim committed:
- Исправил ошибку.

EntityVadim committed:
- Сделал скриншот-тесты.

EntityVadim committed:
- Просил коммит.

EntityVadim committed:
- Исправил key для локализации;
- Дополнил локализацию;
- Организовал код.

EntityVadim committed:
- Просил коммит.

EntityVadim committed:
- Добавил градиентную рамку и два лейбла внутрь.

EntityVadim committed:
- Лейбл сделал поверх маски для верного отображения;
- Настроил все по макету.

EntityVadim committed:
- Написал метод, который обновляет статистику, как только завершу трекер.

EntityVadim committed:
- Исправил критические ошибки по обновлении статистики.

EntityVadim committed:
- Улучшил удаление трекеров.

EntityVadim committed:
- Переделал логику с категорией "Закрепленные".

EntityVadim committed:
- Организовал код, для лучшей читаемости.

EntityVadim committed:
- Просил коммит.

EntityVadim committed:
- Подключил аналитику.

EntityVadim committed:
- Исправил баг с фильтром (некорректно отображался завершенный трекер);
- Организовал код.

protos17 commented on Podfile.lock:
Если присмотреться, то тут новых 135 тысяч строк))) Чтобы такого не было, файлы pods не нужно отправлять в репозиторий, они подтягиваются самостоятельно. Достаточно только оставить файл `Podfile` (файл `Podfile.lock` тоже не нужен). Для того, чтобы git самостоятельно эти файлы не добавля, необходимо в репозиторий добавить файл `.gitignore`. Обычно достаточно стандартного, можешь взять его отсюда: https://github.com/github/gitignore/blob/main/Swift.gitignore
но там раскомментируй `Pods`.
Вот что написано об этом в `Cocoapods`: https://guides.cocoapods.org/using/using-cocoapods.html#should-i-check-the-pods-directory-into-source-control

protos17 commented on Tracker/AppMetrica/AppMetrica.swift:
```diff
@@ -0,0 +1,49 @@
+    case track = "track"
+    case filter = "filter"
+    case edit = "edit"
+    case delete = "delete"
```
Если кейс enum'a не отличается от его rawValue, то можно не писать его:

      case addTrack = "add_track"
      case track
      case filter

protos17 commented on Tracker/AppMetrica/AppMetrica.swift:
```diff
@@ -0,0 +1,49 @@
+
+struct AppMetricaCore {
+    static func activate() {
+        guard let configuration = AppMetricaConfiguration(apiKey: "e04b57c6-b7ad-4740-b897-e5c90606bf1e")
```
apiKey рекомендую хранить в константах

protos17 commented on Tracker/Statistics/StatisticsViewController.swift:
```diff
@@ -61,6 +126,20 @@ final class StatisticsViewController: UIViewController {
+            gradientView.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -16),
+            gradientView.heightAnchor.constraint(equalToConstant: 90),
+            
+            counterLabel.topAnchor.constraint(equalTo: gradientView.topAnchor, constant: 12),
+            counterLabel.leadingAnchor.constraint(equalTo: gradientView.leadingAnchor, constant: 12),
```
Константы для констрейнтов рекомендую выносить в extension класса: 

       extension StatisticsViewController {          
            private enum Constants {          
                  static let defaultOffset: CGFloat = 12              
                     ...            
                    }
              }

protos17 commented on Tracker/Statistics/StatisticsViewController.swift:
```diff
@@ -74,10 +153,20 @@ final class StatisticsViewController: UIViewController {
+            gradientView.isHidden = true
+            errorImageView.isHidden = false
+            placeholderLabel.isHidden = false
+            counterLabel.isHidden = true
```
Чтобы не дублировать, предлагаю использовать тернарный оператор:

      gradientView.isHidden = completedTrackersCount < 0

protos17 commented on Tracker/Trackers/TrackerFilter/ExtensionTrackerFilter.swift:
```diff
@@ -0,0 +1,74 @@
+        cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+            let cell = tableView.dequeueReusableCell(
+                withIdentifier: TrackerFilterViewController.cellIdentifier,
+                for: indexPath) as! TrackerCategoryCell
```
Не рекомендую использовать принудительное приведение типов (as!), это может привести к вылету приложения

protos17 commented on Tracker/Trackers/TrackerFilter/ExtensionTrackerFilter.swift:
```diff
@@ -0,0 +1,74 @@
+    func tableView(
+        _ tableView: UITableView,
+        heightForRowAt indexPath: IndexPath) -> CGFloat {
+            return 75
```
При возвращении значения в одну строку можно не писать слово return

protos17 commented on Tracker/Trackers/TrackerFilter/TrackerFilterViewController.swift:
```diff
@@ -0,0 +1,219 @@
+    }
+    
+    private func loadSelectedFilter() -> Int? {
+        return UserDefaults.standard.object(forKey: TrackerFilterViewController.selectedFilterKey) as? Int
```
Работу с файлами рекомендую перенести в отдельный менеджер

protos17 commented on Tracker/Trackers/TrackerViewController/TrackerCollection/TrackerCell.swift:
```diff
@@ -201,4 +230,42 @@ final class TrackerCell: UICollectionViewCell {
         delegate?.trackerCellDidToggleCompletion(self, for: tracker)
     }
+    
+    @objc func presentContextMenu() {
```
Здесь стоит добавить `private`
Необходимо указывать private  для того, чтобы спрятать детали реализации конкретной части функциональности, когда они используются только внутри области объявления.

protos17 commented on Tracker/Trackers/TrackerViewController/TrackerCollection/TrackerCell.swift:
```diff
@@ -201,4 +230,42 @@ final class TrackerCell: UICollectionViewCell {
+            let pinActionTitle = isPinned ?
+            NSLocalizedString("Unpin", comment: "Открепить") :
+            NSLocalizedString("Pin", comment: "Закрепить")
+            let pinAction = UIAction(title: pinActionTitle, image: UIImage(systemName: "pin")) { _ in
```
Сильная ссылка, необходимо добавить [weak self]

protos17 commented on Tracker/Trackers/TrackerViewController/TrackerCollection/TrackerCell.swift:
```diff
@@ -201,4 +230,42 @@ final class TrackerCell: UICollectionViewCell {
+                self.delegate?.trackerCellDidToggleCompletion(self, for: tracker)
+            }
+            let editAction = UIAction(
+                title: NSLocalizedString("Edit", comment: "Редактировать")) { _ in
```
Здесь тоже сильная ссылка

protos17 commented on Tracker/Trackers/TrackerViewController/TrackerCollection/TrackerCell.swift:
```diff
@@ -201,4 +230,42 @@ final class TrackerCell: UICollectionViewCell {
+            let deleteAction = UIAction(
+                title: NSLocalizedString("Delete",comment: "Удалить"),
+                attributes: .destructive) { _ in
+                    self.delegate?.trackerCellDidRequestDelete(self, for: tracker)
```
Здесь сильная ссылка

protos17 commented on Tracker/Trackers/TrackerViewController/TrackerViewController.swift:
```diff
@@ -89,41 +117,134 @@ final class TrackerViewController: UIViewController {
+        let deleteAction = UIAlertAction(
+            title: "Удалить",
+            style: .destructive
+        ) { _ in
```
Сильная ссылка

EntityVadim committed:
Удалены файлы Pods и Podfile.lock, добавлен .gitignore

EntityVadim committed:
- Добавил .gitignore.

EntityVadim committed:
- Исправил замечания ревью:
  - Если кейс enum'a не отличается от его rawValue, то можно не писать его;
  - apiKey рекомендую хранить в константах.

EntityVadim committed:
- Добавил private, чтобы спрятать детали реализации.

EntityVadim committed:
- Расширил .gitignore;
- Устранил дубляж кода.

EntityVadim committed:
- Устранил сильные ссылки.

EntityVadim committed:
- Устранил сильные ссылки.

EntityVadim committed:
- Устранил ошибку с локализацией;
- Клавиатура скрывается при вводе текста в поиске.
